### ISSUE TITLE
Diff setting via configuration/environment variable has no effect

### ISSUE THREAD

**Issue by nblock** [ansible/ansible#23912]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
core

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /home/user/code/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.11 (default, Mar 20 2016, 15:46:50) [GCC 5.3.0]
```

##### CONFIGURATION
depending on the scenario, see below:
* OK: when using ```--diff``` command line parameter
* FAIL: when using ```ANSIBLE_DIFF_ALWAYS=true ansible-playbook [args]```
* FAIL: with the following snippet in ansible.cfg:
```
  [diff]
  always = True
```

##### OS / ENVIRONMENT
Arch Linux

##### SUMMARY
The diff setting has no effect when specified via configuration file or environment variable.

##### STEPS TO REPRODUCE
* Create the following playbook:
```
---
- hosts: all
  tasks:
    - name: Set a random number in motd
      template:
        src: nodiff-motd.j2
        dest: /etc/motd
        owner: root
        group: root
        mode: 0644
```
* Create the nodiff-motd.j2 template:
```
The current random number is: {{ 1000 | random }}
```
* A inventory with a single host.

##### EXPECTED RESULTS
A diff output with each run (it should not matter where the diff option is configured).

##### ACTUAL RESULTS
* Works if diff is enabled via command line parameter: ```$ ansible-playbook -i inventories/development.ini --diff nodiff.yml```
```
  TASK [Set a random number in motd] ****
  --- before: /etc/motd
  +++ after: dynamically generated
  @@ -1 +1 @@
  -The current random number is: 257
  +The current random number is: 682
```
* Does not work with the following snippet in ansible.cfg:
```
  [diff]
  always = true
```
```
  $ ansible-playbook -i inventories/development.ini nodiff.yml
  [snipped]
  TASK [Set a random number in motd] ****
  changed: [devel]
```
* Does not work when specified as environment variable: ```$ ANSIBLE_DIFF_ALWAYS=1 ansible-playbook -i inventories/development.ini nodiff.yml```
```
  TASK [Set a random number in motd] ****
  changed: [devel]
```

This feature was added with: https://github.com/ansible/ansible/commit/2291163a7a37876fde2b63746ca52aeba0bb7e2a

**Comment by alikins** [comment]
Currently, ANSIBLE_DIFF_ALWAYS=true  and the config setting are a little different than the cli --diff options.

ANSIBLE_DIFF_ALWAYS=true and config option makes tasks/modules always provide diff information in their returned data. 

Adding '--diff' to the cli also does that, but also shows that data from the tasks (via the default callback,  but the diff callback is only triggered if the the strategy sees the cli '--diff' option via PlaybookExecutor->TaskQueueManager->Strategy).

But that is just current behavior, making them behave the same seems reasonable to me.

**Comment by alikins** [comment]
Looking through the code in devel/ and  2291163 I see a couple of problems that may be involved.

1) the def of ansible.constant.DIFF_ALWAYS uses the wrong value_type ('bool' should be 'boolean')

2) ansible.plugins.strategy.Strategy.parent_handler_match only calls the diff callback if the --diff cli option is provided and ignores DIFF_ALWAYS

**Comment by samdoran** [comment]
This should be fixed in newer versions of Ansible. If you are still experiencing this problem, please open a new issue.
